src/ExportHistory/Activities/ExportInstanceHistoryActivity.cs

```diff
@@ -0,0 +1,321 @@
+    static string GetFileExtension(ExportFormat format)
+    {
+        return format.Kind switch
+        {
+            ExportFormatKind.Jsonl => "jsonl.gz",  // JSONL format is compressed
+            ExportFormatKind.Json => "json",       // JSON format is uncompressed
+            _ => "jsonl.gz",                       // Default to JSONL compressed
+        };
+    }
+
+    static string SerializeInstanceData(
+        List<HistoryEvent> historyEvents,
+        ExportFormat format)
+    {
+        JsonSerializerOptions serializerOptions = new JsonSerializerOptions
+        {
+            WriteIndented = false,
+            DefaultIgnoreCondition = JsonIgnoreCondition.WhenWritingNull,
+            PropertyNamingPolicy = JsonNamingPolicy.CamelCase,
+            IncludeFields = true, // Include fields, not just properties (matches JsonDataConverter pattern)
+            Converters = { new JsonStringEnumConverter() }, // Serialize enums as strings
+        };
+
+        if (format.Kind == ExportFormatKind.Jsonl)
+        {
+            // JSONL format: one history event per line
+            // Serialize as object to preserve runtime type (polymorphic serialization)
+            StringBuilder jsonlBuilder = new();
+
+            foreach (HistoryEvent historyEvent in historyEvents)
+            {
+                // Serialize as object to preserve the actual derived type
+                string json = JsonSerializer.Serialize((object)historyEvent, historyEvent.GetType(), serializerOptions);
+                jsonlBuilder.AppendLine(json);
+            }
+
+            return jsonlBuilder.ToString();
+        }
+        else
+        {
+            // JSON format: array of history events
+            // Convert to object array to preserve runtime types
+            object[] eventsAsObjects = historyEvents.Cast<object>().ToArray();
+            return JsonSerializer.Serialize(eventsAsObjects, serializerOptions);
+        }
+    }
+
+    async Task UploadToBlobStorageAsync(
+        string containerName,
+        string blobPath,
+        string content,
+        ExportFormat format,
+        string instanceId,
+        CancellationToken cancellationToken)
+    {
+        // Create blob service client from connection string
+        // Note: Azure.Storage.Blobs clients (BlobServiceClient, BlobContainerClient, BlobClient) are lightweight
+        // wrappers that don't implement IDisposable/IAsyncDisposable. They use HttpClient internally which is
+        // managed by the framework and doesn't require explicit disposal. The clients are designed to be
+        // stateless and safe for reuse or short-lived usage without disposal.
+        BlobServiceClient serviceClient = new(this.storageOptions.ConnectionString);
+        BlobContainerClient containerClient = serviceClient.GetBlobContainerClient(containerName);
+
+        // Ensure container exists
+        await containerClient.CreateIfNotExistsAsync(
```

Comment:
Is there anyway we can do this in an earlier step? Checking the container exists every upload is a lot of extra operations.